File: python/problem46.py
import math


def goldbachOtherConjecture():
    isPrime = [True for i in range(6000)]

    for i in range(2, 6000 // 2):
        j = 2
        while (i * j < 6000):
            isPrime[i * j] = False
            j += 1

    compNumbers = []
    primeNumbers = []
    for i in range(2, 6000):
        if isPrime[i]:
            primeNumbers.append(i)
        else:
            compNumbers.append(i)

    for i in range(0, len(compNumbers)):
        if compNumbers[i] % 2 == 0: continue
        if not isGoldbach(compNumbers[i], primeNumbers):
            return compNumbers[i]

    return -1


def isGoldbach(num, primeList):
    
    for i in range(0, len(primeList)):
        if (primeList[i] > num): return False
        temp = num - primeList[i]
        if temp % 2 != 0:
            continue
        sqrtVal = math.sqrt(temp // 2)
        if sqrtVal.is_integer():
            return True

    return False

if __name__=="__main__":
    print(goldbachOtherConjecture())

File: go/problem46.go
package main

import (
	"fmt"
	"math"
)

func main() {
	fmt.Println(GoldbachConjecture(6000))
}

func GoldbachConjecture(limit int) int {

	composites, primes := sieveComposite(limit)

	for i := 0; i < len(composites); i++ {
		if composites[i]%2 == 0 {
			continue // skip even composite numbers
		}
		primeSubList := primes[:indexOf(composites[i], primes)]
		if !isGoldbach(composites[i], primeSubList) {
			return composites[i]
		}
	}

	return -1 // Not found
}

// Primes less than a given composite number is constructed
// element is the composite to check
func indexOf(element int, intList []int) int {
	for k, v := range intList {
		if v > element {
			return k
		}
	}
	return len(intList) //For not found, return the whole list
}

func isGoldbach(num int, primeList []int) bool {

	temp := 0
	for i := 0; i < len(primeList); i++ {
		temp = num - primeList[i]
		if temp%2 != 0 {
			continue
		}
		sqrtVal := math.Sqrt(float64(temp / 2))
		if sqrtVal == math.Trunc(sqrtVal) {
			return true
		}
	}
	return false

}

func sieveComposite(limit int) ([]int, []int) {

	if limit <= 1 {
		return []int{}, []int{}
	}

	isPrime := make(map[int]bool)
	for i := 2; i <= limit; i++ {
		isPrime[i] = true
	}

	for i := 2; i <= limit/2; i++ {
		for j := 2; i*j <= limit; j++ {
			isPrime[i*j] = false
		}
	}

	var compNumbers []int
	var primeNumbers []int
	for i := 2; i <= limit; i++ {
		if !isPrime[i] {
			compNumbers = append(compNumbers, i)
		} else {
			primeNumbers = append(primeNumbers, i)
		}
	}

	return compNumbers, primeNumbers
}
